content/blog/taming-react-re-renders.md

```diff
@@ -0,0 +1,174 @@
+
+Custom hooks are great for reusing logic, but they can cause performance issues behind your back. Here's an example:
+
+```jsx
+function useWindowWidth() {
+  const [width, setWidth] = useState(window.innerWidth);
+  useEffect(() => {
+    const onResize = () => setWidth(window.innerWidth);
+    window.addEventListener('resize', onResize);
+    return () => window.removeEventListener('resize', onResize);
+  }, []);
+  return width;
+}
+
+function Header() {
+  const width = useWindowWidth();
+  console.log('Header render');
+  return <h1>Window: {width}px</h1>;
+}
+```
+
+Every time you resize your browser window, even by just 1 pixel, the `Header` component re-renders! If you use this hook in multiple components, you could end up with a lot of unnecessary re-renders.
+
+To avoid this:
+
+- Use throttling or debouncing for frequent updates
+- Only use hooks in components that really need them
+
+## The "Big Re-renders" Myth
+
+Many developers worry about "big" components causing performance issues. But here's the truth: React's diffing algorithm is very efficient! Even lists with dozens of items render quickly.
+
+{{< figure src="/images/2025/taming-react-re-renders/props-myth.jpg" caption="" >}}
+
+Instead of worrying about component size, focus on:
+
+- Optimizing parent components
+- Stabilizing prop references
+- Using memoization when it makes sense
+
+Here's how to use memoization:
+
+```jsx
+const MemoizedList = React.memo(function ({items}) {
+  console.log('List render');
+  return null;
+});
+
+// In parent:
+const stableItems = useMemo(() => items, [items]);
+<MemoizedList items={stableItems} />;
+```
+
+## The solution: Moving State Down
+
+One of the best ways to prevent unnecessary re-renders is to move state as close as possible to where it's used. Here's an example:
+
+```jsx
+// Before: parent holds all item-states
+function Parent({initialItems}) {
+  const [items, setItems] = useState(initialItems);
+  return items.map((item, i) => (
+    <Item
+      key={i}
+      item={item}
+      onUpdate={(newItem) => {
+        const copy = [...items];
+        copy[i] = newItem;
+        setItems(copy);
+      }}
+    />
+  ));
+}
+
+// After: each Item manages its own state
+function Parent({initialItems}) {
+  return initialItems.map((item, i) => <Item key={i} initial={item} />);
+}
+
+function Item({initial}) {
+  const [item, setItem] = useState(initial);
+  return (
+    <div>
+      <input
+        value={item.text}
+        onChange={(e) => setItem({...item, text: e.target.value})}
+      />
+    </div>
+  );
+}
+```
+
+By moving state down to individual `Item` components, typing in one input only re-renders that specific item, not the entire list!
+
+Remember: Optimizing React performance is more about understanding when and why components re-render than about complex optimizations. Start by profiling your app with React DevTools, identify the bottlenecks, and apply these strategies where they make the most sense.
```

Comment:
todo: lets add a link to react devtools here, also this is another great tool for catching re-renders: https://github.com/welldone-software/why-did-you-render you can add this too